### PR TITLE
Validator ordering changes

### PR DESCRIPTION
**Pull request by piotr-roslaniec**
**Type of PR:**
- Other

**Required reviews:** 
- 1

**What this does:**
- Replaces global validator ordering with canonical share indices in DKG instances
- Replaces `DkgValidator` with `Validator`
- Prevents panic when aggregating empty PVSS instances list
- Deduplicates share utils/variables
- Fixes some internal TODOs
- Removes unused code

**Issues fixed/closed:**
> - Fixes #...

**Why it's needed:**
- Removes some technical debt and lays the groundwork for other, functional issues

**Notes for reviewers:**
- There's a lot of smaller changes that make understanding this PR a bit difficult. I've added comments in the PR itself as pointers to more important pieces.

### CODE REVIEW

File: ferveo/src/validator.rs [hunk 1]
```diff
@@ -45,29 +47,36 @@ pub struct Validator<E: Pairing> {
...
-    }
-}
-
-impl<E: Pairing> Ord for Validator<E> {
-    // Validators are ordered by their address only
-    fn cmp(&self, other: &Self) -> Ordering {
-        self.address.cmp(&other.address)
-    }
+    /// The index of the validator in the given ritual
+    pub share_index: u32,
```
Comment (piotr-roslaniec) [inline ferveo/src/validator.rs:51]: DKG expects to be provided with a `share_index` for every `Validator`

File: ferveo/src/bindings_python.rs [hunk 2]
```diff
@@ -109,6 +108,14 @@ impl From<FerveoPythonError> for PyErr {
                         "num_shares: {num_shares}, security_threshold: {security_threshold}"
                     ))
                 },
+                Error::DuplicatedShareIndex(index) => {
```
Comment (piotr-roslaniec) [inline ferveo/src/bindings_python.rs:111]: New exceptions

File: ferveo/src/api.rs [hunk 3]
```diff
@@ -637,6 +658,41 @@ mod test_ferveo_api {
         assert!(local_aggregate
             .verify(dkg.0.dkg_params.shares_num(), &messages)
             .is_ok());
+
+        // Test negative cases
```
Comment (piotr-roslaniec) [inline ferveo/src/api.rs:662]: These test cases were adapted from the server variant of tests above

File: ferveo/src/validator.rs [hunk 4]
```diff
@@ -45,29 +47,36 @@ pub struct Validator<E: Pairing> {
...
     ) -> Result<Self, EthereumAddressParseError> {
         Ok(Self {
             address: EthereumAddress::from_str(&address)?,
             public_key,
+            share_index,
         })
     }
 }
+
+pub fn assert_no_share_duplicates<E: Pairing>(
```
Comment (piotr-roslaniec) [inline ferveo/src/validator.rs:68]: Replaces validation logic for ensuring validator ordering

**Review by cygnusv — APPROVED** [review]

File: ferveo/src/api.rs [hunk 5]
```diff
@@ -222,10 +222,14 @@ impl Dkg {
...
-        &self,
+        &mut self,
         rng: &mut R,
     ) -> Result<Transcript> {
-        self.0.create_share(rng)
+        match self.0.share(rng) {
+            Ok(Message::Deal(transcript)) => Ok(transcript),
+            Err(e) => Err(e),
+            _ => Err(Error::InvalidDkgStateToDeal),
+        }
```
[the remark this answers is not in the document]
Comment (piotr-roslaniec) [reply ferveo/src/api.rs:233]: It's a pattern-matching statement with different cases on each line: 
- The first line describes a case where we get the intended result - We just need to unwrap `transcript` from a `Message` struct, and specifically from `Message::Deal` variant. I'm considering deprecating `Message` as we only use it internally.
- The second line handles the error result
- The third one handles other `Message` variants. There are two currently, `Message::Deal` and `Message:Aggregate`. We treat results other than `Message::Deal` as errors, as we don't expect the DKG to be ready to aggregate at this point.

**Pull request merged**